modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/metadata/ColocationGroup.java

```diff
@@ -228,6 +228,27 @@ public ColocationGroup finalaze() {
         return forNodes0(nodeIds);
     }
 
+    /** */
+    public ColocationGroup filterForPartitions(int[] parts) {
+        if (assignments != null) {
+            List<List<UUID>> assignments = new ArrayList<>(this.assignments.size());
+            Set<UUID> nodes = new HashSet<>();
+
+            for (int i = 0; i < this.assignments.size(); ++i) {
+                UUID first = F.isEmpty(parts) || Arrays.binarySearch(parts, i) >= 0 ? F.first(this.assignments.get(i)) : null;
```

Comment:
1. `F.isEmpty(parts)` can be checked once (together with assignment check)
2. We can't be sure that `parts` is sorted, this variable is passed by user. Perhaps it's better to build BitSet and use it to filter partitions.